# Projecto 1 

### Integrantes

* Jonathan Rivera
* Julian Castro  
* Alejandro Gomez

### Caso

El Ministerio de Comercio, Industria y Turismo de Colombia, la Asociación Hotelera y
Turística de Colombia – COTELCO, cadenas hoteleras de la talla de Hilton, Hoteles Estelar,
Holiday Inn y hoteles pequeños ubicados en diferentes municipios de Colombia están
interesados en analizar las características de sitios turísticos que los hacen atractivos para
turistas locales o de otros países, ya sea para ir a conocerlos o recomendarlos. De igual
manera, quieren comparar las características de dichos sitios, con aquellos que han
obtenido bajas recomendaciones y que están afectando el número de turistas que llegan a
ellos. Adicionalmente, quieren tener un mecanismo para determinar la calificación que
tendrá un sitio por parte de los turistas y así, por ejemplo, aplicar estrategias para identificar
oportunidades de mejora que permitan aumentar la popularidad de los sitios y fomentar el
turismo.
Esos actores de turismo prepararon dos conjuntos de datos con reseñas de sitios turísticos.
Cada reseña tiene una calificación según el sentimiento que tuvo el turista al visitarlo. Estos
actores quieren lograr un análisis independiente de los conjuntos de datos y al final del
proyecto discutir sobre los grupos de científicos de datos e ingenieros de datos que
acompañarán el desarrollo real de este proyecto.

### Objetivos

Plantear un modelo de prediccion de calificaiciones de reseñas, basado en el procesamiento de textos, con un algoritmo que identifique la relacion semantica entre palabras y las calificaciones numericas

In [1]:
%pip install nltk
%pip install openpyxl
%pip install contractions
%pip install inflect
%pip install spacy
%pip install gensim
%pip install scikit-optimize
!python -m spacy download es_core_news_md
%pip install inflect
%pip install scikit-plot
%pip install spacy-language-detection
%pip install -U pip setuptools wheel
%pip install -U spacy
!python -m spacy download en_core_web_sm
! pip install langdetect

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
  Attempting uninstall: typing-extensions
    Found existing installation: typing-extensions 4.9.0
    Uninstalling typing-extensions-4.9.0:
      Successfully uninstalled typing-extensions-4.9.0
Note: you may need to restart the kernel to use updated packages.
  Attempting uninstall: colorama
    Found existing installation: colorama 0.4.4
    Uninstalling colorama-0.4.4:
      Successfully uninstalled colorama-0.4.4
Note: you may need to restart the kernel to use updated packages.
  Created wheel for fst-pso: filename=fst_pso-1.8.1-py3-none-any.whl size=20445 sha256=f6cef033ca9b89856c606e402f2c06e716c922165e2a866e9238dd7ff1522404
  Stored in directory: c:\users\julia\appdata\local\pip\cache\wheels\99\66\48\d7ce0c6927f6abf167bbcdee537affc7b92c03632f78028411
  Created wheel for miniful: file

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
spyder 5.1.5 requires pyqt5<5.13, which is not installed.
spyder 5.1.5 requires pyqtwebengine<5.13, which is not installed.
conda-repo-cli 1.0.4 requires pathlib, which is not installed.
anaconda-project 0.10.1 requires ruamel-yaml, which is not installed.



  Attempting uninstall: wheel
    Found existing installation: wheel 0.37.0
    Uninstalling wheel-0.37.0:
      Successfully uninstalled wheel-0.37.0
  Attempting uninstall: setuptools
    Found existing installation: setuptools 58.0.4
    Uninstalling setuptools-58.0.4:
      Successfully uninstalled setuptools-58.0.4
  Attempting uninstall: pip
    Found existing installation: pip 21.2.4
    Uninstalling pip-21.2.4:
      Successfully uninstalled pip-21.2.4
Note: you may need to restart the kernel to use updated packages.


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


     ---------------------------------------- 12.8/12.8 MB 1.0 MB/s eta 0:00:00

DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063



✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')

DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


### Instalacion de librerias

In [3]:
import nltk
# Punkt permite separar un texto en frases.
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download("averaged_perceptron_tagger") 


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\julia\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\julia\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\julia\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\julia\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [4]:
import pandas as pd
import numpy as np
import sys
!{sys.executable} -m pip install pandas-profiling

import re, string, unicodedata
import contractions
import inflect
from nltk import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import LancasterStemmer, WordNetLemmatizer

from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer, HashingVectorizer
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.svm import SVC
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import BernoulliNB
from sklearn.metrics import classification_report, confusion_matrix
from nltk import word_tokenize, sent_tokenize
from sklearn.metrics import ConfusionMatrixDisplay, RocCurveDisplay, roc_auc_score, precision_score, recall_score, f1_score
from sklearn.linear_model import LogisticRegression


from sklearn.base import BaseEstimator, ClassifierMixin

import matplotlib.pyplot as plt

     -------------------------------------- 85.1/85.1 kB 680.4 kB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
     -------------------------------------- 61.0/61.0 kB 806.3 kB/s eta 0:00:00
INFO: pip is looking at multiple versions of visions to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of visions[type-image-path] to determine which version is compatible with other requirements. This could take a while.
   -------------------------------------- 324.4/324.4 kB 957.7 kB/s eta 0:00:00
   -------------------------------------- 357.9/357.9 kB 968.4 kB/s eta 0:00:00
   -------------------------------------- 296.5/296.5 kB 917.6 kB/s eta 0:00:00
   ---------------------------------------- 2.6/2.6 MB 976.1 kB/s eta 0:00:00
   -------------------------------------- 666.3/666.3 kB 976.3 kB/s eta 0:00:00
   --------------------------------

DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063
ERROR: Cannot uninstall 'llvmlite'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.


## 1. Perfilamiento y entendimiento de los datos

In [6]:
data=pd.read_csv('tipo2_entrenamiento_estudiantes.csv', sep=',', encoding = 'utf-8')
# Asignación a una nueva variable de los datos leidos
data_t=data

In [7]:
data.head()

Review  Class
0  Muy buena atención y aclaración de dudas por p...      5
1  Buen hotel si están obligados a estar cerca de...      3
2  Es un lugar muy lindo para fotografías, visite...      5
3  Abusados con la factura de alimentos siempre s...      3
4  Tuvimos un par de personas en el grupo que rea...      3

Los datos que tenemos se presentan de una manera muy simple, una columna con las reseñas en lenguaje natural, y otra columna con la calificación numerica de las reseñas

### 1.1 Lenguaje de los datos

Para el planteamiento de un modelo que analice textos, es esencial primero conocer el idioma en el que se encuentran los textos, esto porque tanto para la transformación de los textos (lematización, tokenización) para poder ser procesado por el modelo, como para configurar el modelo, es necesario que se tenga un lenguaje comun para todos los tokens

In [8]:
#Podemos usar la libreria langdetect para conocer el idioma de todos los textos, y podemos aprovecharnos de las funcionalidades 
# de pandas para facilmente agregar una columna con el idioma del texto para su posterior analizis
from langdetect import detect
data_t['idioma'] = data_t['Review'].apply(detect)
data_t

Review  Class idioma
0     Muy buena atención y aclaración de dudas por p...      5     es
1     Buen hotel si están obligados a estar cerca de...      3     es
2     Es un lugar muy lindo para fotografías, visite...      5     es
3     Abusados con la factura de alimentos siempre s...      3     es
4     Tuvimos un par de personas en el grupo que rea...      3     es
...                                                 ...    ...    ...
7870  Me parece buen sistema, agiliza el transporte,...      4     es
7871  Fue una escapada de un día desde el complejo, ...      4     es
7872  La Plaza de la Revolución es un lugar emblemát...      3     es
7873  Es la segunda ocasión que me quedo en los cuar...      1     es
7874  Llegamos por casualidad a Los Mercaderes, un g...      5     es

[7875 rows x 3 columns]

In [9]:
#Podemos realizar el conteo de los valores en la columna idioma para entender la proporcion linguistica de las reseñas
data_t['idioma'].value_counts()

es    7866
en       3
it       3
pt       2
sq       1
Name: idioma, dtype: int64

Como podemos virtualmente todas las reseñas se encuentran en español, con tan solo 8 reseñas de 7874 perteneciendo a otro idioma.
Como mencionabamos previamente, el modelo tiene que plantarse para un solo idioma, por lo que sera necesario remover las reseñas de otros idiomas en la fase de transformación de los datos

### 1.2 Entendimiento de los ratings

Entender y analizar la distribución de los datos de los rating puede ayudarnos a contribuir cualitativamente los objetivos de negocios

In [10]:
data_t.describe()

Class
count  7875.000000
mean      3.502603
std       1.320435
min       1.000000
25%       3.000000
50%       4.000000
75%       5.000000
max       5.000000

A partir de las estadisticas de los ratings de las reseñas podemos entender lo siguiente:
* No hay ratings invalidos dado que el maximo y el minimo se encuentran en una escala logica y consistente (1-5)
* En promedio las reseñas tienen una calificación de 3.5, por lo que puede ser evidencia de que se tiene una distribución apropiada de calificaciones (no estan sesgadas a ningun extremo).
* Lo anterior se puede confirmar observando la distribución de los datos en los cuartiles. El 25% de los ratings tienen una calificación maxima de 3, el 50% de 4, y el 75% de 5.
* Tenemos 7875 reseñas para la creación del modelo y su posterior validación. 

### 1.3 Calidad de los datos

#### 1.3.1 Completitud

In [11]:
((data_t.isnull().sum()/data_t.shape[0])).sort_values(ascending=False)

Review    0.0
Class     0.0
idioma    0.0
dtype: float64

Podemos observar que no tenemos valores nulos en nuestros datos, por lo tanto no sera necesario hacer un analizis de manejo de nulalidad

#### 1.3.2 Unicidad

In [12]:
data_t.duplicated(keep = False).sum()

102

Podemos observar que tenemos 102 registros duplicados, debido a que los duplicados no añaden valor alguno al modelo, pero si pueden introducir un sesgo (debido a que ciertas palabras pueden asosiarse de manera mas fuerte a un rating de manera erronea), tendremos que eliminarlos en la fase de limpieza de datos.

#### 1.3.3 Consistencia

Como veiamos previamente en el entendimiento de los ratings, estos datos son consistentes dado que todas las calificaciones se encuentran en un rango coherente y logico de (1-5)

#### 1.3.4 Validez

Debido a que no existe una relación cuantificable entre las columnas, no tiene cabida un analisis de validez de las relaciones entre columnas

### 2. Preparación de los datos

#### 2.1 Limpieza de los datos 

#### 2.1.1 Quitar duplicados

In [13]:
data_t= data_t.drop_duplicates()
data_t.duplicated(keep = False).sum()

0

Como mencionamos previamente, tenemos que deshacernos de los duplicados antes de alimentar el modelo con los datos. Observamos que el dataframe queda sin datos duplicados

#### 2.1.2 Quitar reseñas en otros idiomas

In [14]:
data_t= data_t[data_t.idioma == 'es']
data_t['idioma'].value_counts()

es    7793
Name: idioma, dtype: int64

Como se menciono en el entendimiento de datos, nos vamos a quedar con solo las reseñas que esten en español. Podemos observar que el dataframe ahora solo contiene reseñas en español

#### 2.1.3 Transformar los datos

In [17]:
stop_words = stopwords.words("spanish")
def remove_non_ascii(words):
    """Remove non-ASCII characters from list of tokenized words"""
    new_words = []
    for word in words:
        new_word = unicodedata.normalize('NFKD', word).encode('ascii', 'ignore').decode('utf-8', 'ignore')
        new_words.append(new_word)
    return new_words
def to_lowercase(words):
    new_words = []
    for word in words:
        new_words.append(word.lower())
    return new_words
    """Convert all characters to lowercase from list of tokenized words"""
def remove_punctuation(words):
    """Remove punctuation from list of tokenized words"""
    new_words = []
    for word in words:
        new_word = re.sub(r'[^\w\s]', '', word)
        if new_word != '':
            new_words.append(new_word)
    return new_words
def replace_numbers(words):
    """Replace all interger occurrences in list of tokenized words with textual representation"""
    p = inflect.engine()
    new_words = []
    for word in words:
        if word.isdigit():
            new_word = p.number_to_words(word)
            new_words.append(new_word)
        else:
            new_words.append(word)
    return new_words
def remove_stopwords(words):
    stop_words = set(stopwords.words("spanish"))
    new_words = [word for word in words if word not in stop_words]
    return new_words
def preprocessing(words):
    words = to_lowercase(words)
    words = replace_numbers(words)
    words = remove_punctuation(words)
    words = remove_non_ascii(words)
    words = remove_stopwords(words)
    return words
preprocessing()

TypeError: preprocessing() missing 1 required positional argument: 'words'

#### Vectorización con TF-IDF

In [18]:
tfidf = TfidfVectorizer(tokenizer=word_tokenize, stop_words=stop_words, lowercase=True)

In [ ]:
X_tfidf = tfidf.fit_transform(X_train["Textos_espanol"])

In [ ]:
print("Vocabulary size:", len(tfidf.vocabulary_))

#### TF-IDF: Clasificación con Regresión Logística

In [ ]:
logistic_model = LogisticRegression(random_state=5, max_iter=1000)

In [ ]:
logistic_model.fit(X_tfidf, y_train)

In [ ]:
y_test_logistic_predict = logistic_model.predict(tfidf.transform(X_test["Textos_espanol"]))

In [ ]:
ConfusionMatrixDisplay.from_predictions(y_test, y_test_logistic_predict)

In [ ]:
precision_logistic = precision_score(y_test, y_test_logistic_predict, average='weighted')
recall_logistic = recall_score(y_test, y_test_logistic_predict, average='weighted')
f1_logistic = f1_score(y_test, y_test_logistic_predict, average='weighted')

print("Precision:", precision_logistic)
print("Recall:", recall_logistic)
print("F1:", f1_logistic)

#### TF-IDF: Clasificación con Random Forest

In [ ]:
tfidf_model = RandomForestClassifier(random_state=3)

In [ ]:
tfidf_model.fit(X_tfidf, y_train)

In [ ]:
pd.Series(tfidf_model.feature_importances_, index=tfidf.vocabulary_).sort_values().tail(20).plot.barh(figsize=(15, 10))

In [ ]:
tfidf_estimators = tfidf_model.estimators_
print("Number of trees:", len(tfidf_estimators))
print("Trees depth (mean):", np.mean([tree.get_depth() for tree in tfidf_estimators]))

In [ ]:
y_train_tfidf_predict = tfidf_model.predict(X_tfidf)
y_test_tfidf_predict = tfidf_model.predict(tfidf.transform(X_test["Textos_espanol"]))

In [ ]:
ConfusionMatrixDisplay.from_predictions(y_train, y_train_tfidf_predict)

In [ ]:
ConfusionMatrixDisplay.from_predictions(y_test, y_test_tfidf_predict)

In [ ]:
print("Precision:", precision_score(y_train, y_train_tfidf_predict,average='weighted'))
print("Recall:", recall_score(y_train, y_train_tfidf_predict,average='weighted'))
print("F1:", f1_score(y_train, y_train_tfidf_predict,average='weighted'))

In [ ]:
print("Precision:", precision_score(y_test, y_test_tfidf_predict,average='weighted'))
print("Recall:", recall_score(y_test, y_test_tfidf_predict,average='weighted'))
print("F1:", f1_score(y_test, y_test_tfidf_predict,average='weighted'))

#### TF-IDF: Clasifiación con SVM

In [ ]:
svm_model = SVC(kernel='linear', random_state=5)

In [ ]:
svm_model.fit(X_tfidf, y_train)

In [ ]:
y_test_svm_predict = svm_model.predict(tfidf.transform(X_test["Textos_espanol"]))

In [ ]:
ConfusionMatrixDisplay.from_predictions(y_test, y_test_svm_predict)

In [ ]:
precision_svm = precision_score(y_test, y_test_svm_predict, average='weighted')
recall_svm = recall_score(y_test, y_test_svm_predict, average='weighted')
f1_svm = f1_score(y_test, y_test_svm_predict, average='weighted')

print("Precision (SVM):", precision_svm)
print("Recall (SVM):", recall_svm)
print("F1 (SVM):", f1_svm)